arithmetization/src/main/java/net/consensys/linea/sequencer/txselection/selectors/TraceLineLimitTransactionSelector.java

```diff
@@ -0,0 +1,100 @@
+import org.hyperledger.besu.plugin.data.TransactionProcessingResult;
+import org.hyperledger.besu.plugin.data.TransactionSelectionResult;
+import org.hyperledger.besu.plugin.services.tracer.BlockAwareOperationTracer;
+import org.hyperledger.besu.plugin.services.txselection.PluginTransactionSelector;
+
+/**
+ * This class implements TransactionSelector and provides a specific implementation for evaluating
+ * transactions based on the number of trace lines per module created by a transaction. It checks if
+ * adding a transaction to the block pushes the trace lines for a module over the limit.
+ */
+@Slf4j
+public class TraceLineLimitTransactionSelector implements PluginTransactionSelector {
+
+  private final Supplier<Map<String, Integer>> moduleLimitsProvider;
+  private final ZkTracer zkTracer;
+  private final String limitFilePath;
+
+  public TraceLineLimitTransactionSelector(
+      final Supplier<Map<String, Integer>> moduleLimitsProvider, final String limitFilePath) {
+    this.moduleLimitsProvider = moduleLimitsProvider;
+    zkTracer = new ZkTracer();
+    zkTracer.traceStartConflation(1L);
+    this.limitFilePath = limitFilePath;
+  }
+
+  /**
+   * No checking is done pre-processing.
+   *
+   * @param pendingTransaction The transaction to evaluate.
+   * @return TransactionSelectionResult.SELECTED
+   */
+  @Override
+  public TransactionSelectionResult evaluateTransactionPreProcessing(
+      final PendingTransaction pendingTransaction) {
+    return TransactionSelectionResult.SELECTED;
+  }
+
+  @Override
+  public void onTransactionNotSelected(
+      final PendingTransaction pendingTransaction,
+      final TransactionSelectionResult transactionSelectionResult) {
+    zkTracer.popTransaction(pendingTransaction);
+  }
+
+  /**
+   * Checking the created trace lines is performed post-processing.
+   *
+   * @param pendingTransaction The processed transaction.
+   * @param processingResult The result of the transaction processing.
+   * @return BLOCK_FULL if the trace lines for a module are over the limit, otherwise SELECTED.
+   */
+  @Override
+  public TransactionSelectionResult evaluateTransactionPostProcessing(
+      final PendingTransaction pendingTransaction,
+      final TransactionProcessingResult processingResult) {
+    final Map<String, Integer> moduleLimits = moduleLimitsProvider.get();
+    // check that we are not exceed line number for any module
+    final Map<String, Integer> lineCounts = zkTracer.getModulesLineCount();
+    for (var e : lineCounts.entrySet()) {
+      final String module = e.getKey();
+      if (!moduleLimits.containsKey(module)) {
+        final String errorMsg =
+            "Module " + module + " does not exist in the limits file: " + limitFilePath;
+        log.error(errorMsg);
+        throw new RuntimeException(errorMsg);
+      }
+      if (lineCounts.get(module) > moduleLimits.get(module)) {
+        return TransactionSelectionResult.BLOCK_FULL;
```

Comment:
do we need to give any more details? like which module limit was exceeded?

Reply:
does it matter which order the module limits are checked?

Reply:
As far as I understand any of the modules over the limit means that the block is full. We could log something if that was helpful ...

Reply:
Order should not matter, AFAIU